openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIResponseMapper.java

```diff
@@ -780,6 +787,75 @@ private ApiResponses createResponseForRecord(String referenceName, Map<String, S
         return Optional.of(apiResponses);
     }
 
+    /**
+     * This function is to handle headers which come with response.
+     *
+     * @param headers   Headers already generated for cache config
+     * @param header    Response header field
+     */
+    private void extractResponseHeaders(Map<String, Header> headers, RecordFieldSymbol header) {
+
+        if (header.typeDescriptor().typeKind() == TypeDescKind.TYPE_REFERENCE) {
+            TypeReferenceTypeSymbol headerType = (TypeReferenceTypeSymbol) header.typeDescriptor();
+            if (headerType.typeDescriptor() instanceof RecordTypeSymbol) {
+                RecordTypeSymbol recordTypeSymbol = (RecordTypeSymbol) headerType.typeDescriptor();
+                Map<String, RecordFieldSymbol> rfields = recordTypeSymbol.fieldDescriptors();
+                Map<String, Header> responseHeaders = new HashMap<>();
+                for (Map.Entry<String, RecordFieldSymbol> field: rfields.entrySet()) {
+                    Header headerSchema = new Header();
+                    TypeSymbol fieldType = field.getValue().typeDescriptor();
+                    String type = fieldType.typeKind().toString().trim().toLowerCase(Locale.ENGLISH);
+                    Schema openApiSchema = ConverterCommonUtils.getOpenApiSchema(type);
+                    if (fieldType instanceof ArrayTypeSymbol) {
+                        ArrayTypeSymbol array = (ArrayTypeSymbol) fieldType;
+                        TypeSymbol itemType = array.memberTypeDescriptor();
+                        String item = itemType.typeKind().toString().trim().toLowerCase(Locale.ENGLISH);
+                        Schema<?> itemSchema = ConverterCommonUtils.getOpenApiSchema(item);
+                        if (openApiSchema instanceof ArraySchema) {
+                            ((ArraySchema) openApiSchema).setItems(itemSchema);
+                        }
+                    }
+                    headerSchema.setSchema(openApiSchema);
+                    Optional<Documentation> fieldDoc = ((Documentable) field.getValue()).documentation();
+                    if (fieldDoc.isPresent() && fieldDoc.get().description().isPresent()) {
+                        headerSchema.setDescription(fieldDoc.get().description().get().trim());
+                    }
+                    // Normalized header key values
+                    String headerKey = getValidHeaderKey(field);
+                    responseHeaders.put(headerKey, headerSchema);
+                }
+                headers.putAll(responseHeaders);
+            }
+        }
+    }
+
+    /**
+     * This function to normalized header key value to OAS key value.
+     * ex: ballerina header key x\-rate\-limit\-id -> X-Rate-Limit-Id
+     * @param field header flied
```

Comment:
Can we expect any other special characters here apart from `-`?

Reply:
After discussing the name represent we decided to go with what the user provides as a name and only change the escape character there `\`. I think it will solve other characters as well. WDYT 😃 ?

Reply:
Yeah, that approach will be better IMO. +1